File: backend/src/odyssey/intent/reentry.py
"""Guilt-free deterministic re-entry policy after absence."""

from datetime import timedelta
from enum import StrEnum
from typing import Literal

from pydantic import AwareDatetime, Field, model_validator

from odyssey.domain.common import UUID7, StrictModel


class ReentryOption(StrEnum):
    CONTINUE = "continue"
    REVISE_SEASON = "revise season"
    STAY_QUIET = "stay quiet"


class ReentryReason(StrEnum):
    ABSENCE_THRESHOLD_MET = "ABSENCE_THRESHOLD_MET"
    MATERIAL_CHANGES_SUMMARIZED = "MATERIAL_CHANGES_SUMMARIZED"
    STALE_OPPORTUNITIES_EXPIRED = "STALE_OPPORTUNITIES_EXPIRED"
    ONE_CLARIFICATION_SELECTED = "ONE_CLARIFICATION_SELECTED"
    NO_CURRENT_MATERIAL_CHANGE = "NO_CURRENT_MATERIAL_CHANGE"


class MaterialChange(StrictModel):
    id: UUID7
    occurred_at: AwareDatetime
    summary: str = Field(min_length=1, max_length=500)
    relevance: float = Field(ge=0, le=1)
    material: bool = True
    currently_relevant: bool = True
    unresolved: bool = False
    clarification_question: str | None = Field(default=None, min_length=1, max_length=500)
    clarification_value: float = Field(default=0, ge=0, le=1)
    source_refs: tuple[UUID7, ...] = ()

    @model_validator(mode="after")
    def validate_clarification(self) -> "MaterialChange":
        if self.clarification_question is None and self.clarification_value != 0:
            raise ValueError("clarification value requires a clarification question")
        return self


class ReentryOpportunity(StrictModel):
    id: UUID7
    expires_at: AwareDatetime
    active: bool = True


class ReentrySummaryItem(StrictModel):
    change_id: UUID7
    summary: str
    source_refs: tuple[UUID7, ...]


class ReentryPolicy(StrictModel):
    minimum_absence: timedelta = Field(default=timedelta(days=3), gt=timedelta(0))
    minimum_material_relevance: float = Field(default=0.5, ge=0, le=1)
    maximum_summary_items: int = Field(default=3, ge=1, le=3)
    policy_version: str = "reentry-policy-1.0"


class ReentrySurface(StrictModel):
    summary: tuple[ReentrySummaryItem, ...] = Field(max_length=3)
    one_question: str | None = None
    options: tuple[ReentryOption, ...]
    expired_opportunity_ids: tuple[UUID7, ...]
    suppress_backlog: Literal[True] = True
    no_absence_penalty: Literal[True] = True
    reason_codes: tuple[ReentryReason, ...]
    policy_version: str


def should_enter_reentry(
    *,
    last_seen: AwareDatetime | None,
    now: AwareDatetime,
    policy: ReentryPolicy | None = None,
) -> bool:
    if last_seen is None:
        return False
    if last_seen > now:
        raise ValueError("last_seen cannot be in the future")
    active_policy = policy or ReentryPolicy()
    return now - last_seen >= active_policy.minimum_absence


def build_reentry(
    *,
    last_seen: AwareDatetime,
    now: AwareDatetime,
    changes: tuple[MaterialChange, ...],
    opportunities: tuple[ReentryOpportunity, ...],
    policy: ReentryPolicy | None = None,
) -> ReentrySurface:
    if last_seen > now:
        raise ValueError("last_seen cannot be in the future")
    active_policy = policy or ReentryPolicy()
    eligible_changes = tuple(
        change
        for change in changes
        if last_seen < change.occurred_at <= now
        and change.material
        and change.currently_relevant
        and change.relevance >= active_policy.minimum_material_relevance
    )
    ranked_changes = sorted(
        eligible_changes,
        key=lambda change: (-change.relevance, -change.occurred_at.timestamp(), str(change.id)),
    )
    summary = tuple(
        ReentrySummaryItem(
            change_id=change.id,
            summary=change.summary,
            source_refs=change.source_refs,
        )
        for change in ranked_changes[: active_policy.maximum_summary_items]
    )
    unresolved = sorted(
        (
            change
            for change in eligible_changes
            if change.unresolved and change.clarification_question is not None
        ),
        key=lambda change: (
            -change.clarification_value,
            -change.relevance,
            -change.occurred_at.timestamp(),
            str(change.id),
        ),
    )
    question = unresolved[0].clarification_question if unresolved else None
    expired_ids = tuple(
        sorted(
            (
                opportunity.id
                for opportunity in opportunities
                if opportunity.active and opportunity.expires_at <= now
            ),
            key=str,
        )
    )
    reasons = [ReentryReason.ABSENCE_THRESHOLD_MET]
    if summary:
        reasons.append(ReentryReason.MATERIAL_CHANGES_SUMMARIZED)
    else:
        reasons.append(ReentryReason.NO_CURRENT_MATERIAL_CHANGE)
    if expired_ids:
        reasons.append(ReentryReason.STALE_OPPORTUNITIES_EXPIRED)
    if question is not None:
        reasons.append(ReentryReason.ONE_CLARIFICATION_SELECTED)
    return ReentrySurface(
        summary=summary,
        one_question=question,
        options=(
            ReentryOption.CONTINUE,
            ReentryOption.REVISE_SEASON,
            ReentryOption.STAY_QUIET,
        ),
        expired_opportunity_ids=expired_ids,
        reason_codes=tuple(reasons),
        policy_version=active_policy.policy_version,
    )

File: apple/Packages/OdysseyIntelligence/Sources/OdysseyIntelligence/Reentry.swift
import Foundation
import OdysseyDomain

public enum ReentryError: Error, Equatable, Sendable {
    case invalidClock
    case invalidText
    case invalidRelevance
    case invalidPolicy
    case futureLastSeen
}

public enum ReentryOption: String, Codable, CaseIterable, Hashable, Sendable {
    case `continue`
    case reviseSeason = "revise_season"
    case stayQuiet = "stay_quiet"
}

public enum ReentryReason: String, Codable, Hashable, Sendable {
    case absenceThresholdMet = "ABSENCE_THRESHOLD_MET"
    case materialChangesSummarized = "MATERIAL_CHANGES_SUMMARIZED"
    case staleOpportunitiesExpired = "STALE_OPPORTUNITIES_EXPIRED"
    case oneClarificationSelected = "ONE_CLARIFICATION_SELECTED"
    case noCurrentMaterialChange = "NO_CURRENT_MATERIAL_CHANGE"
}

public struct ReentryMaterialChange: Codable, Hashable, Sendable {
    public let id: UUIDv7
    public let occurredAt: Date
    public let summary: String
    public let relevance: Double
    public let isMaterial: Bool
    public let isCurrentlyRelevant: Bool
    public let isUnresolved: Bool
    public let clarificationQuestion: String?
    public let clarificationValue: Double
    public let sourceReferences: [UUIDv7]

    public init(
        id: UUIDv7 = UUIDv7(),
        occurredAt: Date,
        summary: String,
        relevance: Double,
        isMaterial: Bool = true,
        isCurrentlyRelevant: Bool = true,
        isUnresolved: Bool = false,
        clarificationQuestion: String? = nil,
        clarificationValue: Double = 0,
        sourceReferences: [UUIDv7] = []
    ) throws {
        guard occurredAt.timeIntervalSinceReferenceDate.isFinite else {
            throw ReentryError.invalidClock
        }
        guard Self.validText(summary, maximum: 500),
              Self.validOptionalText(clarificationQuestion, maximum: 500)
        else {
            throw ReentryError.invalidText
        }
        guard relevance.isFinite,
              (0 ... 1).contains(relevance),
              clarificationValue.isFinite,
              (0 ... 1).contains(clarificationValue),
              clarificationQuestion != nil || clarificationValue == 0
        else {
            throw ReentryError.invalidRelevance
        }
        self.id = id
        self.occurredAt = occurredAt
        self.summary = summary
        self.relevance = relevance
        self.isMaterial = isMaterial
        self.isCurrentlyRelevant = isCurrentlyRelevant
        self.isUnresolved = isUnresolved
        self.clarificationQuestion = clarificationQuestion
        self.clarificationValue = clarificationValue
        self.sourceReferences = sourceReferences
    }

    private static func validText(_ value: String, maximum: Int) -> Bool {
        (1 ... maximum).contains(value.count)
            && value == value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func validOptionalText(_ value: String?, maximum: Int) -> Bool {
        guard let value else { return true }
        return validText(value, maximum: maximum)
    }
}

public struct ReentryOpportunity: Codable, Hashable, Sendable {
    public let id: UUIDv7
    public let expiresAt: Date
    public let isActive: Bool

    public init(
        id: UUIDv7 = UUIDv7(),
        expiresAt: Date,
        isActive: Bool = true
    ) throws {
        guard expiresAt.timeIntervalSinceReferenceDate.isFinite else {
            throw ReentryError.invalidClock
        }
        self.id = id
        self.expiresAt = expiresAt
        self.isActive = isActive
    }
}

public struct ReentryPolicy: Codable, Hashable, Sendable {
    public let minimumAbsence: TimeInterval
    public let minimumMaterialRelevance: Double
    public let maximumSummaryItems: Int
    public let policyVersion: String

    public init(
        minimumAbsence: TimeInterval = 3 * 24 * 60 * 60,
        minimumMaterialRelevance: Double = 0.5,
        maximumSummaryItems: Int = 3,
        policyVersion: String = "reentry-policy-1.0"
    ) throws {
        guard minimumAbsence.isFinite,
              minimumAbsence > 0,
              minimumMaterialRelevance.isFinite,
              (0 ... 1).contains(minimumMaterialRelevance),
              (1 ... 3).contains(maximumSummaryItems),
              (1 ... 100).contains(policyVersion.count),
              policyVersion
                == policyVersion.trimmingCharacters(in: .whitespacesAndNewlines)
        else {
            throw ReentryError.invalidPolicy
        }
        self.minimumAbsence = minimumAbsence
        self.minimumMaterialRelevance = minimumMaterialRelevance
        self.maximumSummaryItems = maximumSummaryItems
        self.policyVersion = policyVersion
    }
}

public struct ReentrySummaryItem: Codable, Hashable, Sendable {
    public let changeID: UUIDv7
    public let summary: String
    public let sourceReferences: [UUIDv7]
}

public struct ReentrySurface: Codable, Hashable, Sendable {
    public let summary: [ReentrySummaryItem]
    public let oneQuestion: String?
    public let options: [ReentryOption]
    public let expiredOpportunityIDs: [UUIDv7]
    public let suppressBacklog: Bool
    public let noAbsencePenalty: Bool
    public let reasons: [ReentryReason]
    public let policyVersion: String
}

public struct ReentryProjector: Sendable {
    public init() {}

    public func shouldEnter(
        lastSeen: Date?,
        now: Date
    ) throws -> Bool {
        try shouldEnter(
            lastSeen: lastSeen,
            now: now,
            policy: ReentryPolicy()
        )
    }

    public func shouldEnter(
        lastSeen: Date?,
        now: Date,
        policy: ReentryPolicy
    ) throws -> Bool {
        guard now.timeIntervalSinceReferenceDate.isFinite else {
            throw ReentryError.invalidClock
        }
        guard let lastSeen else { return false }
        guard lastSeen.timeIntervalSinceReferenceDate.isFinite else {
            throw ReentryError.invalidClock
        }
        guard lastSeen <= now else { throw ReentryError.futureLastSeen }
        return now.timeIntervalSince(lastSeen) >= policy.minimumAbsence
    }

    public func project(
        lastSeen: Date,
        now: Date,
        changes: [ReentryMaterialChange],
        opportunities: [ReentryOpportunity]
    ) throws -> ReentrySurface {
        try project(
            lastSeen: lastSeen,
            now: now,
            changes: changes,
            opportunities: opportunities,
            policy: ReentryPolicy()
        )
    }

    public func project(
        lastSeen: Date,
        now: Date,
        changes: [ReentryMaterialChange],
        opportunities: [ReentryOpportunity],
        policy: ReentryPolicy
    ) throws -> ReentrySurface {
        guard try shouldEnter(lastSeen: lastSeen, now: now, policy: policy) else {
            throw ReentryError.invalidPolicy
        }
        let eligible = changes.filter {
            $0.occurredAt > lastSeen
                && $0.occurredAt <= now
                && $0.isMaterial
                && $0.isCurrentlyRelevant
                && $0.relevance >= policy.minimumMaterialRelevance
        }.sorted(by: rankChanges)
        let summary = eligible.prefix(policy.maximumSummaryItems).map {
            ReentrySummaryItem(
                changeID: $0.id,
                summary: $0.summary,
                sourceReferences: $0.sourceReferences
            )
        }
        let unresolved = eligible.filter {
            $0.isUnresolved && $0.clarificationQuestion != nil
        }.sorted {
            if $0.clarificationValue != $1.clarificationValue {
                return $0.clarificationValue > $1.clarificationValue
            }
            return rankChanges($0, $1)
        }
        let question = unresolved.first?.clarificationQuestion
        let expired = opportunities.filter {
            $0.isActive && $0.expiresAt <= now
        }.map(\.id).sorted { $0.description < $1.description }
        var reasons: [ReentryReason] = [.absenceThresholdMet]
        reasons.append(summary.isEmpty
            ? .noCurrentMaterialChange
            : .materialChangesSummarized)
        if !expired.isEmpty { reasons.append(.staleOpportunitiesExpired) }
        if question != nil { reasons.append(.oneClarificationSelected) }
        return ReentrySurface(
            summary: summary,
            oneQuestion: question,
            options: [.continue, .reviseSeason, .stayQuiet],
            expiredOpportunityIDs: expired,
            suppressBacklog: true,
            noAbsencePenalty: true,
            reasons: reasons,
            policyVersion: policy.policyVersion
        )
    }

    private func rankChanges(
        _ left: ReentryMaterialChange,
        _ right: ReentryMaterialChange
    ) -> Bool {
        if left.relevance != right.relevance { return left.relevance > right.relevance }
        if left.occurredAt != right.occurredAt { return left.occurredAt > right.occurredAt }
        return left.id.description < right.id.description
    }
}
